Description of the Erigon service (#59)
golemfactory/yagna-docs: requestor-tutorials/service-development/service-example-2-managed-erigon.md

---

> description: Ethereum client as a service, implemented as a self-contained runtime

Reviewer: just had a thought that it's kinda awkward to use "as a" twice in the same sentence ... so maybe e.g. "using" ? or "with the use of" ... 

```suggestion
description: Ethereum client as a service, implemented using a self-contained runtime
```

Reviewer: I'm not entirely convinced to this suggestion ... just couldn't think of anything better quickly

---

# Service Example 2: Managed Erigon

The purpose of this example is to demonstrate building a Golem service using a dedicated, self-contained runtime. It features an Ethereum network client, Erigon (f.k.a. Turbo-geth) started on provider nodes on user's demand. Interaction with the service, once it's running and its location and credentials have been passed to the requestor, is *not* facilitated by Golem though.

{% hint style="info" %}

> This example illustrates following Golem features & aspects:

Reviewer: ```suggestion
This example illustrates the following Golem features & aspects:
```

* Dedicated service runtime implemented with [ya-runtime-sdk](https://github.com/golemfactory/ya-runtime-sdk)
* Service management using [yapapi-service-manager](https://github.com/golemfactory/yapapi-service-manager)
{% endhint %}

{% hint style="info" %}
All code for this example can be found in this repository: [https://github.com/golemfactory/yagna-service-erigon/](https://github.com/golemfactory/yagna-service-erigon/)
{% endhint %}

## Architecture overview

![](erigon_architecture.jpg)

> The application consist of three major components:

Reviewer: ```suggestion
The application consists of three major components:
```

1. **User interface**  

>    A browser-based interface written in TypeScript using [React](https://reactjs.org/). It allows the end user to interact with the application, i.e. start an Erigon node after authenticating with their [Metamask](https://metamask.io/) wallet, and later manage their node(s).

Reviewer: unsure, but I think that "manage" should be in the infinite form to match "allows the user. ... " 

```suggestion
   A browser-based interface written in TypeScript using [React](https://reactjs.org/). It allows the end user to interact with the application, i.e. start an Erigon node after authenticating with their [Metamask](https://metamask.io/) wallet, and later to manage their node(s).
```

2. **Requestor server**  
   A Python application developed using [Quart](https://pgjones.gitlab.io/quart/) and [yapapi-service-manager](https://github.com/golemfactory/yapapi-service-manager). It acts both as an HTTP server (handling requests from the user interface), and as a requestor agent (submitting tasks to the Golem network).
3. **Erigon runtime**  

>     A dedicated, self-contained runtime created with [ya-runtime-sdk](https://github.com/golemfactory/ya-runtime-sdk). It is a Rust binary wrapping [Erigon](https://github.com/ledgerwatch/erigon) service itself so that it can be orchestrated by yagna daemon. The runtime also controls the access to the service by managing [Nginx](https://www.nginx.com/) configuration.

Reviewer: ```suggestion
    A dedicated, self-contained runtime created with [ya-runtime-sdk](https://github.com/golemfactory/ya-runtime-sdk). It is a Rust binary wrapping [Erigon](https://github.com/ledgerwatch/erigon) service itself so that it can be orchestrated by the yagna daemon. The runtime also controls the access to the service by managing [Nginx](https://www.nginx.com/) configuration.
```

   
A typical interaction flow proceeds in the following manner:

1. User opens the web interface in their browser.
2. User authenticates with their metamask wallet.
3. User clicks to start an Erigon node.

> 4. Request to start an Erigon node is sent via REST API to the requestor proxy server.

Reviewer 2: Few reasons why I'm against calling requestor a `proxy`:
* response is returned before any request was even sent to the provider
* `getInstances` is processed on the requestor side only
* there is a lot of communication between `yapapi` and the provider that is not caused by and requests sent to requestor

Author: OK, we can leave it as 'requestor server'.

Author: It was called 'proxy' because the end user doesn't interact directly with the Golem Network. But I'm fine with dropping this term if it suggests that requestor server is doing 1-to-1 proxying of REST API calls to the yagna daemon.

> 4. Request to start an Erigon node is sent via REST API to the requestor server.

Reviewer: ```suggestion
4. Request to start an Erigon node is sent via the REST API to the requestor server.
```

5. Requestor server makes a number of API calls to the yagna daemon running on the requestor's machine in order to find an available Erigon provider and sign an agreement.
6. Once the agreement is signed, requestor server sends a command to the provider to start an Erigon node.
7. Yagna daemon running on the provider's machine spawns Erigon runtime process (a service wrapper/supervisor process).

> 8. Erigon runtime starts the actual Erigon service (backend & rpcdaemon processes).

Reviewer 2: And when does the `nginx` start?

Author: AFAIK it's just running in the background for the whole time. It starts right after it's installed and never stops. @pnowosie correct me if I'm wrong.

Reviewer 2: I think it would be nice to mention this somewhere (I don't know where exactly).
This is far from obvious imho.

9. Erigon runtime generates a new username and password and puts them in the NGINX configuration.

> 10. NGINX auto-reloads on configuration change, from now on granting access to the Erigon service only for the newly created user.

Reviewer 2: I don't like the "newly created user". There is no user, only username-password pair. Maybe 
```
, from now on restricting access to the Erigon service only to requests with these credentials.
```

> 9. Erigon runtime generates a new username and password and puts them in the NGINX configuration.

Reviewer: the numbering is still off here ... ?

Author: Please see my reply on the original comment.

Reviewer: ah, gotcha

11. The credentials are passed backed to the requestor server along with provider's node's public IP address and the port on which NGINX is listening.
12. Requestor server passes all the received information to the web interface where it can be viewed by the user.
13. User interacts with the Erigon service (via NGINX proxy).
14. Once the user finishes using their Erigon instance, they click to stop it.

> 15. Web interface sends request to stop the instance via REST API to the requestor proxy server.

Reviewer 2: requestor `proxy` again - check comment on p.4.

> 15. Web interface sends a request to stop the instance via REST API to the requestor server.

Reviewer: ```suggestion
15. Web interface sends a request to stop the instance via the REST API to the requestor server.
```

16. Requestor server sends a stop command to the provider.
17. Erigon runtime stops the Erigon service and exits.

> 18. Requestor server orchestrates payment for the service in the background without further user interaction.

Reviewer: ```suggestion
18. Requestor server orchestrates a payment for the service in the background without further user interaction.
```

---
description: Ethereum client as a service, implemented as a self-contained runtime
---

# Service Example 2: Managed Erigon

> The purpose of this example is to demonstrate building a Golem service using a dedicated, self-contained runtime. It features an Ethereum network client, Erigon (f.k.a. Turbo-geth) started on provider nodes on user's demand. Interaction with the service is *not* facilitated by Golem though.

Reviewer: ```suggestion
The purpose of this example is to demonstrate building a Golem service using a dedicated, self-contained runtime. It features an Ethereum network client, Erigon (f.k.a. Turbo-geth) started on provider nodes on user's demand. Interaction with the service, once it's running and its location and credentials have been passed to the ruquestor is *not* facilitated by Golem though.
```

{% hint style="info" %}
This example illustrates following Golem features & aspects:

* Dedicated service runtime implemented with [ya-runtime-sdk](https://github.com/golemfactory/ya-runtime-sdk)
* Service management using [yapapi-service-manager](https://github.com/golemfactory/yapapi-service-manager)
{% endhint %}

{% hint style="info" %}
All code for this example can be found in this repository: [https://github.com/golemfactory/yagna-service-erigon/](https://github.com/golemfactory/yagna-service-erigon/)
{% endhint %}

## Architecture overview

![](erigon_architecture.jpg)

The application consist of three major components:

1. **User interface**  
   A browser-based interface written in TypeScript using [React](https://reactjs.org/). It allows the end user to interact with the application, i.e. start an Erigon node after authenticating with their [Metamask](https://metamask.io/) wallet, and later manage their node(s).
2. **Requestor server**  

>    A Python application developed using [Quart](https://pgjones.gitlab.io/quart/) and [yapapi-service-manager](https://github.com/golemfactory/yapapi-service-manager). It acts both as an HTTP server handling requests from the user interface, and as a requestor agent submitting tasks to the Golem network.

Reviewer: ```suggestion
   A Python application developed using [Quart](https://pgjones.gitlab.io/quart/) and [yapapi-service-manager](https://github.com/golemfactory/yapapi-service-manager). It acts both as an HTTP server - handling requests from the user interface, and as a requestor agent - submitting tasks to the Golem network.
```

Author: I'm not sure if that's a correct use of dashes (actually, these are hyphens but never mind that). :thinking:

Author: ```suggestion
   A Python application developed using [Quart](https://pgjones.gitlab.io/quart/) and [yapapi-service-manager](https://github.com/golemfactory/yapapi-service-manager). It acts both as an HTTP server (handling requests from the user interface), and as a requestor agent (submitting tasks to the Golem network).
```

3. **Erigon runtime**  
    A dedicated, self-contained runtime created with [ya-runtime-sdk](https://github.com/golemfactory/ya-runtime-sdk). It is a Rust binary wrapping [Erigon](https://github.com/ledgerwatch/erigon) service itself so that it can be orchestrated by yagna daemon. The runtime also controls the access to the service by managing [Nginx](https://www.nginx.com/) configuration.
   
A typical interaction flow proceeds in the following manner:

1. User opens the web interface in their browser.
2. User authenticates with their metamask wallet.
3. User clicks to start an Erigon node.
4. Request to start an Erigon node is sent via REST API to the requestor server.
5. Requestor starts a service activity (these steps are handled by `yagna` core and `yapapi` internally):
    1. Requestor server makes a number of API calls to the yagna daemon running on the requestor's machine in order to find an available Erigon provider and sign an agreement.
    2. Once the agreement is signed, requestor server sends a command to the provider to start an Erigon node.

>     3. Yagna daemon running on the provider's machine spawns Erigon runtime process (a service wrapper/supervisor process).

Reviewer: ```suggestion
    3. Yagna daemon running on the provider's machine spawns an Erigon runtime process (a service wrapper/supervisor process).
```

8. Erigon runtime starts the actual Erigon service (backend & rpcdaemon processes).
9. Erigon runtime generates a new username and password and puts them in the NGINX configuration.

> 10. NGINX auto-reloads on configuration change, from now on granting access to the Erigon service only for user(s) authenticated with the newly created credentials. (NGINX is continuously running on provider's machine as a daemon, so it doesn't need to be started.)

Reviewer: ```suggestion
10. NGINX auto-reloads on configuration change, from now on granting access to the Erigon service only for user(s) authenticated with the newly-created credentials. (NGINX is continuously running on provider's machine as a daemon, so it doesn't need to be started.)
```

11. The credentials are passed backed to the requestor server along with provider's node's public IP address and the port on which NGINX is listening.
12. Requestor server passes all the received information to the web interface where it can be viewed by the user.
13. User interacts with the Erigon service (via NGINX proxy).
14. Once the user finishes using their Erigon instance, they click to stop it.

> 15. Web interface sends request to stop the instance via REST API to the requestor server.

Reviewer: ```suggestion
15. Web interface sends a request to stop the instance via REST API to the requestor server.
```

> 8. Erigon runtime starts the actual Erigon service (backend & rpcdaemon processes).
> 9. Erigon runtime generates a new username and password and puts them in the NGINX configuration.
> 10. NGINX auto-reloads on configuration change, from now on granting access to the Erigon service only for user(s) authenticated with the newly created credentials. (NGINX is continuously running on provider's machine as a daemon, so it doesn't need to be started.)
> 11. The credentials are passed backed to the requestor server along with provider's node's public IP address and the port on which NGINX is listening.
> 12. Requestor server passes all the received information to the web interface where it can be viewed by the user.
> 13. User interacts with the Erigon service (via NGINX proxy).
> 14. Once the user finishes using their Erigon instance, they click to stop it.
> 15. Web interface sends request to stop the instance via REST API to the requestor server.
> 16. Requestor server sends a stop command to the provider.
> 17. Erigon runtime stops the Erigon service and exits.
> 18. Requestor server orchestrates payment for the service in the background without further user interaction.

Reviewer: shouldn't the numbering here continue from above?:

```suggestion
6. Erigon runtime starts the actual Erigon service (backend & rpcdaemon processes).
7. Erigon runtime generates a new username and password and puts them in the NGINX configuration.
8. NGINX auto-reloads on configuration change, from now on granting access to the Erigon service only for user(s) authenticated with the newly created credentials. (NGINX is continuously running on provider's machine as a daemon, so it doesn't need to be started.)
9. The credentials are passed backed to the requestor server along with provider's node's public IP address and the port on which NGINX is listening.
10. Requestor server passes all the received information to the web interface where it can be viewed by the user.
11. User interacts with the Erigon service (via NGINX proxy).
12. Once the user finishes using their Erigon instance, they click to stop it.
13. Web interface sends request to stop the instance via REST API to the requestor server.
14. Requestor server sends a stop command to the provider.
15. Erigon runtime stops the Erigon service and exits.
16. Requestor server orchestrates payment for the service in the background without further user interaction.
```

... of course, it remains to be seen if the numbering would be preserved by gitbook ...

Author: AFAIK in markdown the actual numbers in the source don't matter. You could even go like this:
1.
1.
1.
...

Author: I'd say: let's merge it like this so we can test if gitbook handles it well.
